odbc/src/api/bind_col_tests.rs

```diff
@@ -0,0 +1,319 @@
+                vec![
+                    MongoColMetadata::new(
+                        "test_db",
+                        "dn".to_string(),
+                        "fn".to_string(),
+                        Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+                        Nullability::SQL_NO_NULLS,
+                        TypeMode::Simple,
+                    ),
+                    MongoColMetadata::new(
+                        "test_db",
+                        "dn".to_string(),
+                        "fn".to_string(),
+                        Schema::Atomic(Atomic::Scalar(BsonTypeName::Int)),
+                        Nullability::SQL_NO_NULLS,
+                        TypeMode::Simple,
+                    ),
+                ],
+            );
+
+            // Must call next to set the `current` field.
+            let _ = mock_query.next(None);
+
+            // Set the mongo_statement
+            *s.mongo_statement.write().unwrap() = Some(Box::new(mock_query.clone()));
+
+            let indicator: *mut Len = null_mut();
+            let buffer: *mut std::ffi::c_void = Box::into_raw(Box::new([0u8; 4])) as *mut _;
+
+            // Assert that SQLBindCol returns an error
+            assert_eq!(
+                SqlReturn::ERROR,
+                SQLBindCol(
+                    stmt as *mut _,
+                    3,
+                    CDataType::SQL_C_SLONG as SmallInt,
+                    buffer,
+                    4,
+                    indicator
+                )
+            );
+
+            // free buffer
+            let _ = Box::from_raw(buffer as *mut WChar);
+        }
+    }
+
+    #[test]
+    fn test_invalid_target_type() {
+        // Set up MongoHandle
+        let env = &mut MongoHandle::Env(Env::with_state(EnvState::Allocated));
+        let conn =
+            &mut MongoHandle::Connection(Connection::with_state(env, ConnectionState::Allocated));
+        let stmt: *mut _ =
+            &mut MongoHandle::Statement(Statement::with_state(conn, StatementState::Allocated));
+
+        unsafe {
+            // Get Statement
+            let s = (*stmt).as_statement().unwrap();
+
+            // set all statement attributes to the correct values.
+            s.attributes.write().unwrap().row_bind_offset_ptr = null_mut();
+            s.attributes.write().unwrap().row_array_size = 1;
+            s.attributes.write().unwrap().row_bind_type = BindType::SQL_BIND_BY_COLUMN as usize;
+
+            // Give bound_cols a column binding with an invalid target_type. The target_type check is in SQLFetch, so if
+            // there is an attempt to create a column binding with an invalid target_type, it will be added to bind_cols,
+            // and SQLBindCol will return SQLReturn::Success. Then, everything will operate as if nothing is wrong until
+            // SQLFetch accesses bound_cols and checks for an invalid target_type. At this point, the error will occur.
```

Comment:
Why is this the case. I see the following in the `SQLBindCol` docs: 

SQLSTATE | Error | Description
-- | -- | --
HY003 | Invalid application buffer type | The argument TargetType was neither a valid data type nor SQL_C_DEFAULT.

Doesn't this mean we should have returned an error from `SQLBindCol` if an invalid target type was provided?

Reply:
Yes you're right. I was thinking It didn't matter where the error was returned as long as it got returned and returning the error in SQLFetch worked smoother with the code; however, I'll go ahead and move it.